google/cloud/security/scanner/audit/fw_rules_engine.py

```diff
@@ -0,0 +1,292 @@
+          policies (list): A list of policy dictionaries.
+
+        Returns:
+          list: A list of FirewallRule.
+        """
+        match_rules = []
+        for policy in policies:
+            rule = firewall_rule.FirewallRule(**policy)
+            match_rules.append(rule)
+        return match_rules
+
+    @property
+    def match_rules(self):
+        """The FirewallRules used to filter policies.
+
+        Returns:
+          list: A list of FirewallRule.
+        """
+        if not self._match_rules:
+            self._match_rules = self.create_rules(self._match_policies)
+        return self._match_rules
+
+    @property
+    def verify_rules(self):
+        """The FirewallRules used to check policies.
+
+        Returns:
+          list: A list of FirewallRule.
+        """
+        if not self._verify_rules:
+            self._verify_rules = self.create_rules(self._verify_policies)
+        return self._verify_rules
+
+    def find_policy_violations(self, firewall_policies):
+        """Finds policy violations in a list of firewall policies.
+
+        Args:
+          firewall_policies (list): A list of FirewallRule.
+
+        Returns:
+          iterable: A generator of RuleViolations.
+        """
+        if self.mode == scanner_rules.RuleMode.MATCHES:
+            violations = self._yield_match_violations(firewall_policies)
+        elif self.mode == scanner_rules.RuleMode.REQUIRED:
+            violations = self._yield_required_violations(firewall_policies)
+        elif self.mode == scanner_rules.RuleMode.WHITELIST:
+            violations = self._yield_whitelist_violations(firewall_policies)
+        elif self.mode == scanner_rules.RuleMode.BLACKLIST:
+            violations = self._yield_blacklist_violations(firewall_policies)
+        return violations
+
+    def _yield_match_violations(self, firewall_policies):
+        """Finds policies that don't match the required policy.
+
+        Args:
+          firewall_policies (list): A list of FirewallRules to check.
+
+        Yields:
+          iterable: A generator of RuleViolations.
+        """
+        recommended_actions = []
+        for rule in self.match_rules:
+            if is_rule_exists_violation(rule, firewall_policies,
+                                        self._exact_match):
+                recommended_action = self._create_recommendation(
+                    'INSERT_FIREWALL_RULE', rule)
+                recommended_actions.append(recommended_action)
+
+        for policy in firewall_policies:
+            if not any(policy == rule for rule in self.match_rules):
+                recommended_action = self._create_recommendation(
+                    'DELETE_FIREWALL_RULE', policy)
+                recommended_actions.append(recommended_action)
+
+        if recommended_actions:
+            yield self._create_violation(
+                firewall_policies, 'FIREWALL_MATCHES_VIOLATION',
+                recommended_actions=[recommended_action])
```

Comment:
Shouldn't this be recommended_actions=recommended_actions? Is there a test that has multiple recommended actions? The test should have failed?

Reply:
Done.